### PR TITLE
JBPM-5902 Stunner - Integration with the jBPM validations API

### PR DESCRIPTION
**Pull request by tiagodolphine**
Now, the **BPMN** validation is applied to the diagram and consolidated with the Graph and Model validations that already existed. The validations alerts keep the same template for messages just appending the new validation messages.
The BPMN validation is based on the **jBPM** implementation in the same way as the old designer does.
It was created a `DomainValidator` interface that represents the validator of the diagram domain, for instance: BPMN, CM, DMN... for now only the BPMN is implemented, but if necessary it is prepared to insert other validators to other domains.
The DomainValidators are executed on backend, while the Graph and Model are still executed on client side.

@romartin 
@LuboTerifaj

### CODE REVIEW

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/validation/BPMNValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+@ApplicationScoped
+public class BPMNValidatorImpl implements BPMNValidator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BPMNValidatorImpl.class);
+    private final DiagramService diagramService;
+    private SemanticModules modules;
+
+    @Inject
+    public BPMNValidatorImpl(final DiagramService diagramService) {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/validation/BPMNValidatorImpl.java]: As this is an `AppScoped` bean, please can you create a no-public, no-arg constructor? That's for the CDI proxy stuff,  
 just for weld not to show the warnings...

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java [hunk 2]
```diff
@@ -94,15 +95,10 @@ private void checkViolations(final H canvasHandler,
...
-        final boolean[] valid = {true};
-        elementViolation.getGraphViolations()
-                .forEach(v -> {
-                    if (applyViolation(canvasHandler,
-                                       v)) {
-                        valid[0] = false;
-                    }
-                });
-        return valid[0];
+        return Stream.concat(elementViolation.getGraphViolations().stream(),
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java]: ohh great change, it was not my best day here by doing the final array trick ehh... :/ well seen, thanks!! :+1:

File: kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-project/pom.xml [hunk 3]
```diff
@@ -607,6 +607,11 @@
       <artifactId>kie-wb-common-stunner-backend-common</artifactId>
     </dependency>
 
+    <dependency>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-project/pom.xml]: this dependency is wrong... lemme explain:
* Stunner provides the common backend stuff in the `kie-wb-common-stunner-backend-common` module, so this module is used in both showcases, np
* On the other hand, when using the project-showcase, it already depends all the `kie-wb-common` and `guvnor` common stuff (from the workbench). So most of the backend stuff that Stunner needs is already inlcuded by other depdencies
* This means that the backend stuff necessary for running the `standalone` and the `project` showcases is different - so the `kie-wb-common-stunner-backend` module is used in the standalone showcase (where no guvnor, etc). And for the project showcase it's being used the `kie-wb-common-stunner-project-backend`. Both are exclusive.. I mean, should not end up in the same classpath... as they provide different implementations for same services

Anyway you didn't really added nothing new in the `kie-wb-common-stunner-backend`, right? So no need for this dependency, isn't it?

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/validation/BPMNViolation.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.Optional;
+
+import org.jboss.errai.common.client.api.annotations.MapsTo;
+import org.jboss.errai.common.client.api.annotations.Portable;
+import org.kie.workbench.common.stunner.core.rule.violations.AbstractRuleViolation;
+import org.kie.workbench.common.stunner.core.validation.DomainViolation;
+
+@Portable
+public class BPMNViolation extends AbstractRuleViolation implements DomainViolation {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/validation/BPMNViolation.java]: why not moving this implementation to some common package? I mean it's nothing concrete for `BPMN` right?  This way once adding new domain validators they can just reuse this class.

**Comment by tiagodolphine** [comment]
Jenkins retest this

**Comment by tiagodolphine** [comment]
hey @romartin I applied your comments.
Thanks !

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/validation/BPMNViolation.java (hunk 4 above)
Comment (romartin) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/validation/BPMNViolation.java]: well yes good point, makes sense too, that's fine for me as it is, np! thanks! :+1:

**Comment by romartin** [comment]
@LuboTerifaj  your turn ;-)

**Comment by LuboTerifaj** [comment]
Hi @romartin
Yes I will have a look at it today or tomorrow.

**Comment by LuboTerifaj** [comment]
Jenkins execute full downstream build

**Comment by LuboTerifaj** [comment]
It looks good, but need to check it in business-central as well. I am waiting for successful full downstream build.

**Comment by LuboTerifaj** [comment]
Hi @tiagodolphine 
I cannot deploy business-central to wildfly. What about you? Have you tried it? Can you deploy it?
Thanks!

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by hasys** [comment]
Hi @LuboTerifaj, @romartin, @manstis, @tiagodolphine, @mareknovotny,

yes, same issues are for other PRs as well. Not sure what is it. @mareknovotny, is it related to some quotas or something like that?

**Comment by mareknovotny** [comment]
@hasys not sure what was there, probably some high load of triggered jobs? Anyway it seems it ran with one failure.

**Comment by tiagodolphine** [comment]
Hey @LuboTerifaj I downloaded the .war and it is not being deployed indeed. I will check why, it seems to be a missing CDI dependency for the `DiagramService` that is necessary on the `BPMNValidadorImpl`.

**Comment by tiagodolphine** [comment]
Hey @LuboTerifaj  @romartin I updated the PR now, to solve the `DiagramService` dependency I created a `DelegateDiagramService` that is an implementation that delegates the call to the `ProjectDiagramService`   on the on the stunner project backend module.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
@LuboTerifaj this is ready to be tested, all green.
Thanks

**Comment by tiagodolphine** [comment]
@romartin just pushed the changing on BPMN Violations to warning. Now the diagram could be saved event if there are any BPMN violations.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Jenkins retest this

**Comment by tiagodolphine** [comment]
@LuboTerifaj  thanks for the review !
@romartin so I think it can be merged now.

**Pull request merged**